File: using_go/session.go
package main

import (
	"errors"
	"fmt"
	"os"
	"strings"
)

// ============================================================
// SESSION MANAGEMENT
// ============================================================
// The session is stored as a plain .session text file at the
// project root. It contains the SHA256 hash of the logged-in
// user's ID. This is intentionally simple — no tokens, no
// expiration, no encryption beyond the hash.
//
// On login:  write SHA256(userID) to .session
// On logout: delete .session
// On any command: read .session, find matching user
// ============================================================

const sessionFile = ".session"

func SaveSession(userID int) error {
	hash := HashSHA256(fmt.Sprintf("%d", userID))
	return os.WriteFile(sessionFile, []byte(hash), 0644)
}

func LoadSession(users []User) (*User, error) {
	data, err := os.ReadFile(sessionFile)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, fmt.Errorf("no active session. Please login first")
		}
		return nil, fmt.Errorf("reading session: %w", err)
	}

	sessionHash := strings.TrimSpace(string(data))

	for i, user := range users {
		userHash := HashSHA256(fmt.Sprintf("%d", user.ID))
		if userHash == sessionHash {
			return &users[i], nil
		}
	}

	return nil, fmt.Errorf("session expired or invalid. Please login again")
}

func ClearSession() error {
	err := os.Remove(sessionFile)
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		return fmt.Errorf("removing session: %w", err)
	}
	return nil
}

File: using_rust/src/session.rs
use std::fs;
use std::io;

use crate::user::{User, hash_sha256};

/// Archivo de sesión que almacena el hash SHA256 del ID del usuario autenticado.
const SESSION_FILE: &str = ".session";

/// Guarda la sesión del usuario escribiendo el hash SHA256 de su ID en el archivo de sesión.
pub fn save_session(user_id: i32) -> Result<(), Box<dyn std::error::Error>> {
    let hash = hash_sha256(&user_id.to_string());
    fs::write(SESSION_FILE, hash)?;
    Ok(())
}

/// Carga la sesión activa comparando el hash almacenado con los IDs de los usuarios registrados.
/// Retorna un `User` clonado si encuentra una coincidencia.
pub fn load_session(users: &[User]) -> Result<User, Box<dyn std::error::Error>> {
    let data = match fs::read_to_string(SESSION_FILE) {
        Ok(content) => content,
        Err(e) if e.kind() == io::ErrorKind::NotFound => {
            return Err("no active session. Please login first".into());
        }
        Err(e) => return Err(format!("reading session: {}", e).into()),
    };

    let session_hash = data.trim();

    for user in users {
        let user_hash = hash_sha256(&user.id.to_string());
        if user_hash == session_hash {
            return Ok(user.clone());
        }
    }

    Err("session expired or invalid. Please login again".into())
}

/// Elimina el archivo de sesión. No retorna error si el archivo no existe.
pub fn clear_session() -> Result<(), Box<dyn std::error::Error>> {
    match fs::remove_file(SESSION_FILE) {
        Ok(()) => Ok(()),
        Err(e) if e.kind() == io::ErrorKind::NotFound => Ok(()),
        Err(e) => Err(format!("removing session: {}", e).into()),
    }
}
